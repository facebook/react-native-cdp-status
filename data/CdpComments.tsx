/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import lineColumn from 'line-column';
import { ReferenceFile, ReferenceLocation } from './ImplementationModel';

type CommentNode = {
  match: string;
  index: number;
  length: number;
  // 1-based
  line: number;
  // 1-based
  column: number;
  cleanedMatch: string;
};

export type CdpComment = ReferenceLocation & {
  cleanedText: string;
  hasAdditionalContent: boolean;
};

function extractCStyleComments(
  code: string,
  tagsOfInterest: string[],
): Array<CommentNode> {
  const lineComment = '\\/\\/[^\n]*';
  const blockComment = '\\/\\*[\\s\\S]*?(?:\\*\\/|$)';
  const lineCommentBlock = `(?:^|\\n)[ \\t]*${lineComment}(?:\\n[ \\t]*${lineComment})*`;
  const inlineComment = '(?:[^\\n]*?)' + lineComment;

  const regex = new RegExp(
    `${lineCommentBlock}|${inlineComment}|${blockComment}`,
    'g',
  );

  const matches: Array<CommentNode> = [];
  const lc = lineColumn(code);
  for (const match of code.matchAll(regex)) {
    // Ignore comments that don't contain any tags of interest
    if (!tagsOfInterest.some((tag) => match[0].includes(tag))) {
      continue;
    }
    const firstNonWhitespaceIndex = match.index! + match[0].search(/\S/);
    const { line, col } = lc.fromIndex(firstNonWhitespaceIndex)!;
    matches.push({
      match: match[0],
      index: match.index!,
      length: match[0].length,
      line,
      column: col,
      cleanedMatch: cleanCommentSource(match[0]),
    });
  }
  return matches;
}

function cleanCommentSource(comment: string) {
  // Remove block comment start/end tokens
  comment = comment
    .replace(/^\/\*\*?/, '')
    .replace(/\*\/$/, '')
    // Remove line comment start tokens
    .replace(/^\s*\/\/\s*/gm, '')
    // Remove leading asterisks from each line
    .replace(/^\s*\* ?/gm, '')
    // Remove leading/trailing newlines
    .replace(/^\n|\n$/g, '')
    // Remove trailing whitespace from all lines
    .replace(/[ \t]+$/gm, '');
  // Find and trim the minimum leading whitespace on any non-empty line
  let minLeadingWhitespace = Infinity;
  const lines = comment.split('\n');
  for (const line of lines) {
    if (line.trim() !== '') {
      const leadingWhitespace = line.match(/^\s*/)![0].length;
      minLeadingWhitespace = Math.min(leadingWhitespace, minLeadingWhitespace);
    }
  }
  // Remove the minimum leading whitespace from each line
  comment = lines.map((line) => line.slice(minLeadingWhitespace)).join('\n');
  return comment;
}

export type ParsedCdpComments = {
  commentsByCdpSymbol: Map<string, Array<CdpComment>>;
};

export function parseCdpComments(code: string): ParsedCdpComments {
  const comments = extractCStyleComments(code, ['@cdp']);
  const commentsByCdpSymbol = new Map<string, Array<CdpComment>>();
  for (const comment of comments) {
    const cdpTagMatches = comment.cleanedMatch.matchAll(/@cdp\s+([\w.]*\w)/g);
    const symbolsSeen = new Set<string>();
    for (const cdpTagMatch of cdpTagMatches) {
      const cdpSymbol = cdpTagMatch[1];
      if (cdpSymbol != null) {
        // Ignore duplicate symbols within the same comment
        if (symbolsSeen.has(cdpSymbol)) {
          continue;
        }
        commentsByCdpSymbol.set(
          cdpSymbol,
          commentsByCdpSymbol.get(cdpSymbol) ?? [],
        );
        commentsByCdpSymbol.get(cdpSymbol)!.push({
          line: comment.line,
          column: comment.column,
          // Further cleaned up version of the comment
          cleanedText: comment.cleanedMatch.replace(/^@cdp\s+/, ''),
          // Is there more to the comment than just the tag+symbol
          hasAdditionalContent:
            comment.cleanedMatch.length > cdpTagMatch[0].length,
        });
      }
    }
  }
  return {
    commentsByCdpSymbol,
  };
}

export type ParsedAndIndexedCdpComments = {
  commentsByCdpSymbol: Map<string, Array<CdpComment & ReferenceFile>>;
};

export function parseAndIndexCdpComments(
  files: Iterable<readonly [ReferenceFile, string]>,
): ParsedAndIndexedCdpComments {
  const commentsByCdpSymbol = new Map<
    string,
    Array<CdpComment & ReferenceFile>
  >();
  for (const [file, code] of files) {
    const comments = parseCdpComments(code);
    for (const [cdpSymbol, cdpComments] of comments.commentsByCdpSymbol) {
      commentsByCdpSymbol.set(
        cdpSymbol,
        commentsByCdpSymbol.get(cdpSymbol) ?? [],
      );
      for (const cdpComment of cdpComments) {
        commentsByCdpSymbol.get(cdpSymbol)!.push({
          ...cdpComment,
          ...file,
        });
      }
    }
  }
  return {
    commentsByCdpSymbol,
  };
}
