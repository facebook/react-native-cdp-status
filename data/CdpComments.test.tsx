/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { parseAndIndexCdpComments } from './CdpComments';

describe('CdpComments', () => {
  const files = [
    [
      {
        github: {
          owner: 'owner',
          repo: 'repo',
          commitSha: '00000000',
          path: 'main.js',
        },
      },
      `
// This is a single line comment mentioning @cdp symbol1
/* @cdp symbol1 is mentioned in this block comment */
/**
 * This is a multiline comment
 * that also mentions @cdp symbol1
 */
const x = 10; // This is an inline @cdp symbol1 comment
const y = 20; // This is another inline @cdp symbol1 comment
/**
 * This is a multiline comment
 * with two lines
 * mentioning @cdp
 * symbol1
 */
// @cdp symbol1
// @cdp symbol2

// separate comment @cdp symbol2

/**
 * @cdp symbol1
 * - embedded markdown
 *   - nested
 */

      /**
       * @cdp Domain.method
       * note that the comment is indented
       */

'@cdp tags outside of comments are ignored';

// @cdp symbol1
'NOTE: the comment above has no additional content'

// @cdp symbol3's mentions can use punctuation without confusing the parser.
// @cdp symbol3.@cdp symbol3, @cdp symbol3?
`,
    ],
  ] as const;

  test('parseAndIndexCdpComments', () => {
    const result = parseAndIndexCdpComments(files);
    expect(result).toMatchSnapshot();
  });
});
