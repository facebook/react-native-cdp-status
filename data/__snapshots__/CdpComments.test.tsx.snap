// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdpComments parseAndIndexCdpComments 1`] = `
{
  "commentsByCdpSymbol": Map {
    "symbol1" => [
      {
        "cleanedText": "This is a single line comment mentioning @cdp symbol1",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 2,
      },
      {
        "cleanedText": "symbol1 is mentioned in this block comment",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 3,
      },
      {
        "cleanedText": "This is a multiline comment
that also mentions @cdp symbol1
",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 4,
      },
      {
        "cleanedText": "const x = 10; // This is an inline @cdp symbol1 comment",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 8,
      },
      {
        "cleanedText": "const y = 20; // This is another inline @cdp symbol1 comment",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 9,
      },
      {
        "cleanedText": "This is a multiline comment
with two lines
mentioning @cdp
symbol1
",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 10,
      },
      {
        "cleanedText": "symbol1
@cdp symbol2",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 16,
      },
      {
        "cleanedText": "symbol1
- embedded markdown
  - nested
",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 21,
      },
      {
        "cleanedText": "symbol1",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": false,
        "line": 34,
      },
    ],
    "symbol2" => [
      {
        "cleanedText": "symbol1
@cdp symbol2",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 16,
      },
      {
        "cleanedText": "separate comment @cdp symbol2",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 19,
      },
    ],
    "Domain.method" => [
      {
        "cleanedText": "Domain.method
note that the comment is indented
",
        "column": 7,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 27,
      },
    ],
    "symbol3" => [
      {
        "cleanedText": "symbol3's mentions can use punctuation without confusing the parser.
@cdp symbol3.@cdp symbol3, @cdp symbol3?",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 37,
      },
      {
        "cleanedText": "symbol3's mentions can use punctuation without confusing the parser.
@cdp symbol3.@cdp symbol3, @cdp symbol3?",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 37,
      },
      {
        "cleanedText": "symbol3's mentions can use punctuation without confusing the parser.
@cdp symbol3.@cdp symbol3, @cdp symbol3?",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 37,
      },
      {
        "cleanedText": "symbol3's mentions can use punctuation without confusing the parser.
@cdp symbol3.@cdp symbol3, @cdp symbol3?",
        "column": 1,
        "github": {
          "commitSha": "00000000",
          "owner": "owner",
          "path": "main.js",
          "repo": "repo",
        },
        "hasAdditionalContent": true,
        "line": 37,
      },
    ],
  },
}
`;
